README.md
## Table of Contents
* [About the project](#about-the-project)
  * [Built With](#built-with)
* [Getting Started](#getting-started)
  * [Prerequisites](#prerequisites)
  * [Installation](#installation)
* [Usage](#usage)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [License](#license)
* [Contact](#contact)
* [Acknowledgements](#acknowledgements)


## About the project
The `AnalysisObserver` observes Landscape Model simulation runs and runs an R script within an R environment whenever
a Monte Carlo run, or an entire experiment, is finished. The observer does also contain a copy of [
FfMpeg](https://ffmpeg.org) for fast rendering of animations.

### Built with
* Landscape Model core v1.6.2
* [R](https://cran.r-project.org) v4.1.1


## Getting Started
The following section gives instructions on how to make use of the `AnalysisObserver` within your model variant. This
information is intended for model developers.

### Prerequisites
Make sure you are using the most recent version of the Landscape Model core that is compatible with this 
`AnalysisObserver` version. The description here assumes a setup as described in the Landscape Model core `README`.

### Installation
1. Copy the observer to the `model\variant` folder of the Landscape Model.
2. Add the observer to the `<Parts>` section of the `model\variant\experiment.xml` file.
3. Provide an R script that will be used for analysis, preferably under `model\variant`. See [usage](#usage) for details
   on how to prepare the R script for running fom the observer.
4. If necessary, install additional required R packages within the R environment bundled with the `AnalysisObserver`. It
   is also possible to update the entire R version, but in this case the path specified in teh component has probably
   also to be updated.
5. Place an `Observer` block within the `model\variant\experiment.xml` and/or the `model\variant\mc.xml` that configures
   the observer to be run after an experiment and/or a Monte Carlo run has finished. See [usage](#usage) for further
   details.


## Usage
The `AnalysisObserver` starts a vanilla instance of the bundled R instance for execution of a specified `Script`. The
observer receives the parameters specified within the observer configuration as named key-value command line 
arguments of the form `--key=value`. Thereby, automatic conversion of keys to lower-case takes place. The R package 
`optparse` can be used for parsing these command line arguments. This makes the configuration of the observer very 
adaptive as the R script developer can specify the arguments needed for the script. The model developer can provide
these arguments through the model configuration. An exception is the two mandatory parameters `Data` and `Output_folder`
that are passed as unnamed arguments in the given order. An example configuration looks like this:

```xml
<Observer module="AnalysisObserver" class="AnalysisObserver">
    <Script>$(_X3DIR_)/../../variant/RiskAnalysis_Experiment.R</Script>
    <Data>$(_EXP_BASE_DIR_)\$(SimID)\mcs</Data>
    <Output_Folder>$(_EXP_BASE_DIR_)\$(SimID)\analysis</Output_Folder>
    <DsName>PecSoil</DsName>
</Observer>
```
If placed in the `<Observers>` section of the `model\variant\experiment.xml`, a script located in the `model\variant`
folder (script path resolved relative to the `_X3DIR_`, see Landscape Model core `README`) is run after the
experiment has finished. `<Data>` (here a folder) and `<Output_folder>` are passed to the script as unnamed command 
line arguments. Additionally, a named argument `dsname` can be accessed from within the script.

The R scripts can make use of the `xRisk` R package that allows accessing X3df data stores.


## Roadmap
The `AnalysisObserver` is considered stable with no further development planned. We suggest to use `ReportingElements`, 
within compositions or within Jupyter notebooks instead.


## Contributing
Contributions are welcome. Please contact the authors (see [Contact](#contact)).


## License
Distributed under the CC0 License. See `LICENSE` for more information.


## Contact
Thorsten Schad - thorsten.schad@bayer.com
Sascha Bub - sascha.bub.ext@bayer.com


## Acknowledgements
* [Apply function progress bars](https://cran.r-project.org/web/packages/pbapply)
* [data.table](https://cran.r-project.org/web/packages/data.table)
* [Direct Labels](https://cran.r-project.org/web/packages/directlabels)
* [FfMpeg](https://ffmpeg.org)
* [ggplot2](https://cran.r-project.org/web/packages/ggplot2)
* [h5](https://cran.r-project.org/web/packages/h5)
* [matrixStats](https://cran.r-project.org/web/packages/matrxistats)
* [openxlsx](https://cran.r-project.org/web/packages/openxlsx)
* [optparse](https://cran.r-project.org/web/packages/optparse)
* [R GDAL bindings](https://cran.r-project.org/web/packages/rgdal)
* [raster](https://cran.r-project.org/web/packages/raster)


CHANGELOG.md
# Changelog

This is the changelog for the AnalysisObserver. It was automatically created on 2023-09-12.

## [2.2.6] - 2023-09-12

### Added

### Changed

- Updated changelog

### Fixed

## [2.2.5] - 2023-09-12

### Added

- Creation of repository info during documentation

- Repository info to R runtime environment

### Changed

- Extended module information for R runtime environment

### Fixed

## [2.2.4] - 2023-09-11

### Added

- Module information

### Changed

### Fixed

## [2.2.3] - 2022-11-01

### Added

- `XRisk` package to runtime environment

### Changed

### Fixed

## [2.2.2] - 2021-12-07

### Added

### Changed

- Spell checking of `README` 

### Fixed

## [2.2.1] - 2021-11-05

### Added

- `IRkernel` to R runtime environment

### Changed

### Fixed

## [2.2] - 2021-10-18

### Added

### Changed

- Updated R packages

- Replaced `rgdal` by `terra` package

### Fixed

## [2.1.3] - 2021-10-15

### Added

### Changed

### Fixed

- Set R module library path

## [2.1.2] - 2021-10-12

### Added

### Changed

- Switched to Google docstring style

### Fixed

## [2.1.1] - 2021-10-11

### Added

### Changed

- Replaced legacy format strings by f-strings

### Fixed

## [2.1] - 2021-09-08

### Added

### Changed

- Updated runtime environment to R 4.1.1

### Fixed

## [2.0.5] - 2021-08-05

### Added

### Changed

- Renamed `LICENSE.txt` to `LICENSE` 

### Fixed

## [2.0.4] - 2021-06-25

### Added

### Changed

- Updated the documentation and started using markdown in changelog

### Fixed

## [2.0.3] - 2021-06-18

### Added

### Changed

- Updated `data.table` package

### Fixed

## [2.0.2] - 2020-12-07

### Added

### Changed

- Line separators in `LICENSE` 

- Changelog and `README` modified

### Fixed

## [2.0.1] - 2020-12-03

### Added

- Changelog and release history

- `README`, `LICENSE`, `CONTRIBUTING` 

### Changed

### Fixed

## [2.0] - 2020-10-22

### Added

### Changed

- First independent release

### Fixed

## [1.3.24] - 2020-04-02

### Added

- `observer.AnalysisObserver`.flush() and observer.AnalysisObserver.write()

### Changed

- `observer.AnalysisObserver` uses base function to call observer module

### Fixed

## [1.3.13] - 2020-02-07

### Added

### Changed

- `observer.AnalysisObserver` no longer throws error if output directory exists

### Fixed

## [1.2.38]

### Added

### Changed

- Update of `observer.AnalysisObserver` `XRisk` package to v0.0.0.9004

### Fixed

## [1.2.36]

### Added

### Changed

- Removed generic R functions from class `observer.AnalysisObserver` 

### Fixed

## [1.2.35]

### Added

### Changed

- Update of `observer.AnalysisObserver` `XRisk` package to v0.0.0.9003

### Fixed

## [1.2.33]

### Added

### Changed

- `observer.AnalysisObserver` no longer requires scripts to be in observer directory

- Scripts in class `observer.AnalysisObserver` directory removed

### Fixed

## [1.2.12]

### Added

- `observer.AnalysisObserver.mc_run_started()` 

### Changed

### Fixed

## [1.2.10]

### Added

### Changed

### Fixed

- Geo-raster orientations in class `observer.AnalysisObserver` 

## [1.2.6]

### Added

### Changed

### Fixed

- `observer.AnalysisObserver` R functions for multiple habitat types

## [1.2.2]

### Added

### Changed

- `observer.AnalysisObserver` do not pass lock keyword to R in `AnalysisObserver` 

### Fixed

## [1.2.1]

### Added

### Changed

- `observer.AnalysisObserver` scripts now use option parser for passing parameters

- `observer.AnalysisObserver` scripts can now analyse data sets other than PEC soil

- `observer.AnalysisObserver` scripts modularized

### Fixed

## [1.1.6]

### Added

### Changed

- `observer.AnalysisObserver` info worksheet prepended to Excel worksheet outputs

### Fixed

## [1.1.3]

### Added

### Changed

- `observer.AnalysisObserver` visualization threshold defined by median value

### Fixed

- `observer.AnalysisObserver` spray-drift/run-off co-occurrence only if both data set exist

## [1.1.2]

### Added

### Changed

### Fixed

- `observer.AnalysisObserver` box-plot labels in `RiskAnalysis_Experiment_PEC` 

## [1.1.1]

### Added

- `observer.AnalysisObserver`

### Changed

### Fixed


document.py
"""
Script for documenting the code of the AnalysisObserver.
"""
import os
import base.documentation
import AnalysisObserver

root_folder = os.path.abspath(os.path.join(os.path.dirname(base.__file__), ".."))
base.documentation.write_changelog(
    "AnalysisObserver",
    AnalysisObserver.AnalysisObserver.VERSION,
    os.path.join(root_folder, "..", "variant", "AnalysisObserver", "CHANGELOG.md")
)
base.documentation.write_contribution_notes(
    os.path.join(root_folder, "..", "variant", "AnalysisObserver", "CONTRIBUTING.md"))
base.documentation.write_repository_info(
    os.path.join(root_folder, "..", "variant", "AnalysisObserver"),
    os.path.join(root_folder, "..", "variant", "AnalysisObserver", "repository.json"),
    os.path.join(root_folder, "..", "..", "..", "versions.json"),
    "observer"
)


AnalysisObserver.py
"""An observer intended to run analysis R scripts during simulation."""
import os
import base


class AnalysisObserver(base.Observer):
    """An observer that runs R scripts."""
    # RELEASES
    VERSION = base.VersionCollection(
        base.VersionInfo("2.2.6", "2023-09-12"),
        base.VersionInfo("2.2.5", "2023-09-12"),
        base.VersionInfo("2.2.4", "2023-09-11"),
        base.VersionInfo("2.2.3", "2022-11-01"),
        base.VersionInfo("2.2.2", "2021-12-07"),
        base.VersionInfo("2.2.1", "2021-11-05"),
        base.VersionInfo("2.2.0", "2021-10-18"),
        base.VersionInfo("2.1.3", "2021-10-15"),
        base.VersionInfo("2.1.2", "2021-10-12"),
        base.VersionInfo("2.1.1", "2021-10-11"),
        base.VersionInfo("2.1.0", "2021-09-08"),
        base.VersionInfo("2.0.5", "2021-08-05"),
        base.VersionInfo("2.0.4", "2021-06-25"),
        base.VersionInfo("2.0.3", "2021-06-18"),
        base.VersionInfo("2.0.2", "2020-12-07"),
        base.VersionInfo("2.0.1", "2020-12-03"),
        base.VersionInfo("2.0.0", "2020-10-22"),
        base.VersionInfo("1.3.24", "2020-04-02"),
        base.VersionInfo("1.3.13", "2020-02-07"),
        base.VersionInfo("1.2.38", None),
        base.VersionInfo("1.2.36", None),
        base.VersionInfo("1.2.35", None),
        base.VersionInfo("1.2.33", None),
        base.VersionInfo("1.2.12", None),
        base.VersionInfo("1.2.10", None),
        base.VersionInfo("1.2.6", None),
        base.VersionInfo("1.2.2", None),
        base.VersionInfo("1.2.1", None),
        base.VersionInfo("1.1.6", None),
        base.VersionInfo("1.1.3", None),
        base.VersionInfo("1.1.2", None),
        base.VersionInfo("1.1.1", None)
    )

    # CHANGELOG
    VERSION.added("1.1.1", "`observer.AnalysisObserver`")
    VERSION.fixed("1.1.2", "`observer.AnalysisObserver` box-plot labels in `RiskAnalysis_Experiment_PEC` ")
    VERSION.fixed("1.1.3", "`observer.AnalysisObserver` spray-drift/run-off co-occurrence only if both data set exist")
    VERSION.changed("1.1.3", "`observer.AnalysisObserver` visualization threshold defined by median value")
    VERSION.changed("1.1.6", "`observer.AnalysisObserver` info worksheet prepended to Excel worksheet outputs")
    VERSION.changed("1.2.1", "`observer.AnalysisObserver` scripts now use option parser for passing parameters")
    VERSION.changed("1.2.1", "`observer.AnalysisObserver` scripts can now analyse data sets other than PEC soil")
    VERSION.changed("1.2.1", "`observer.AnalysisObserver` scripts modularized")
    VERSION.changed("1.2.2", "`observer.AnalysisObserver` do not pass lock keyword to R in `AnalysisObserver` ")
    VERSION.fixed("1.2.6", "`observer.AnalysisObserver` R functions for multiple habitat types")
    VERSION.fixed("1.2.10", "Geo-raster orientations in class `observer.AnalysisObserver` ")
    VERSION.added("1.2.12", "`observer.AnalysisObserver.mc_run_started()` ")
    VERSION.changed("1.2.33", "`observer.AnalysisObserver` no longer requires scripts to be in observer directory")
    VERSION.changed("1.2.33", "Scripts in class `observer.AnalysisObserver` directory removed")
    VERSION.changed("1.2.35", "Update of `observer.AnalysisObserver` `XRisk` package to v0.0.0.9003")
    VERSION.changed("1.2.36", "Removed generic R functions from class `observer.AnalysisObserver` ")
    VERSION.changed("1.2.38", "Update of `observer.AnalysisObserver` `XRisk` package to v0.0.0.9004")
    VERSION.changed("1.3.13", "`observer.AnalysisObserver` no longer throws error if output directory exists")
    VERSION.added("1.3.24", "`observer.AnalysisObserver`.flush() and observer.AnalysisObserver.write()")
    VERSION.changed("1.3.24", "`observer.AnalysisObserver` uses base function to call observer module")
    VERSION.changed("2.0.0", "First independent release")
    VERSION.added("2.0.1", "Changelog and release history")
    VERSION.added("2.0.1", "`README`, `LICENSE`, `CONTRIBUTING` ")
    VERSION.changed("2.0.2", "Line separators in `LICENSE` ")
    VERSION.changed("2.0.2", "Changelog and `README` modified")
    VERSION.changed("2.0.3", "Updated `data.table` package")
    VERSION.changed("2.0.4", "Updated the documentation and started using markdown in changelog")
    VERSION.changed("2.0.5", "Renamed `LICENSE.txt` to `LICENSE` ")
    VERSION.changed("2.1.0", "Updated runtime environment to R 4.1.1")
    VERSION.changed("2.1.1", "Replaced legacy format strings by f-strings")
    VERSION.changed("2.1.2", "Switched to Google docstring style")
    VERSION.fixed("2.1.3", "Set R module library path")
    VERSION.changed("2.2.0", "Updated R packages")
    # noinspection SpellCheckingInspection
    VERSION.changed("2.2.0", "Replaced `rgdal` by `terra` package")
    VERSION.added("2.2.1", "`IRkernel` to R runtime environment")
    VERSION.changed("2.2.2", "Spell checking of `README` ")
    VERSION.added("2.2.3", "`XRisk` package to runtime environment")
    VERSION.added("2.2.4", "Module information")
    VERSION.added("2.2.5", "Creation of repository info during documentation")
    VERSION.added("2.2.5", "Repository info to R runtime environment")
    VERSION.changed("2.2.5", "Extended module information for R runtime environment")
    VERSION.changed("2.2.6", "Updated changelog")

    MODULE = base.Module("R", "4.1.1", "R-4.1.1", "R-4.1.1/README", None, True, "R-4.1.1/doc/NEWS")

    def __init__(self, data, script, output_folder, **keywords):
        """
        Initializes an AnalysisObserver.

        Args:
            data: A reference to the input data to be analyzed.
            script: The file path of the R script run by the AnalysisObserver.
            output_folder: The folder where outputs of the analysis script should be written to.
            **keywords: Additional, script-specific arguments for the analysis.
        """
        super(AnalysisObserver, self).__init__()
        script = os.path.abspath(script)
        if not os.path.isfile(script):
            raise FileNotFoundError(f"AnalysisObserver script '{script}' not found")
        self._componentPath = os.path.dirname(__file__)
        self._wd = os.path.dirname(output_folder)
        # noinspection SpellCheckingInspection
        self._rCall = [os.path.join(self._componentPath, "R-4.1.1", "bin", "x64", "Rscript.exe"), "--vanilla", script]
        for key, value in keywords.items():
            if key != "lock":
                self._rCall.append(f"--{key}={value}")
        self._rCall.extend([data, output_folder])
        os.makedirs(output_folder, exist_ok=True)
        self._env = {"R_LIBS_USER": os.path.join(self._componentPath, "R-4.1.1", "library")}

    def experiment_finished(self, detail=None):
        """
        Reacts when an experiment is completed.

        Args:
            detail: Additional details to report.

        Returns:
             Nothing.
        """
        base.run_process(self._rCall, self._wd, self.default_observer, self._env)

    def input_get_values(self, component_input):
        """
        Reacts when values are requested from a component input.

        Args:
            component_input: The input being requested.

        Returns:
            Nothing.
        """

    def mc_run_finished(self, detail=None):
        """
        Reacts when a Monte Carlo run is finished.

        Args:
            detail: Additional details to report.

        Returns:
             Nothing.
        """
        base.run_process(self._rCall, self._wd, self.default_observer, self._env)
